### ISSUE TITLE
Disallow #[must_use] for functions with no return value

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#55035]
This code:

```
#[must_use]
fn foo() {}
fn main() {
    foo();
}
```

Gives (rustc 1.31.0-nightly 2c2e2c57d 2018-10-12):

```
warning: unused return value of `foo` which must be used
 --> ...\test.rs:4:5
  |
4 |     foo();
  |     ^^^^^^
  |
  = note: #[warn(unused_must_use)] on by default
```

But what I expected is no warning on the foo call, and an error (or warning) for adding #[must_use] on a function without return value.

In (rare) code like this, where the unit (empty tuple) is returned in an explicit way, I expect no errors nor warnings:

```
#[must_use]
fn foo() -> () {}
fn main() {
    let _ = foo();
}
```

**Comment by varkor** [comment]
It was decided `#[must_use]` should act just as on any function in this case: https://github.com/rust-lang/rust/issues/54828. The "explicit return syntax" is semantically no different from the implicit return type and making that different would be misleading.

**Comment by leonardo-m** [comment]
I am not sure that decision is the best, but OK, I close this down.
